simplified/HordeAd/Core/Ast.hs

```diff
@@ -20,30 +23,174 @@ module HordeAd.Core.Ast
+import           Data.Array.Internal (valueOf)
 import qualified Data.Array.RankedS as OR
 import           Data.IORef.Unboxed (Counter, atomicAddCounter_, newCounter)
 import           Data.Kind (Type)
+import           Data.Monoid (All (..))
 import           Data.MonoTraversable (Element, MonoFunctor (omap))
 import qualified Data.Strict.Vector as Data.Vector
 import qualified Data.Vector.Generic as V
 import qualified Data.Vector.Storable as VS
 import           GHC.TypeLits (KnownNat, Nat, type (+))
-import           Numeric.LinearAlgebra (Numeric)
+import           Numeric.LinearAlgebra (Numeric, Vector)
 import           System.IO.Unsafe (unsafePerformIO)
 import           Text.Show.Functions ()
+import           Unsafe.Coerce (unsafeCoerce)
 
 import HordeAd.Internal.OrthotopeOrphanInstances ()
 
--- * Definitions
+-- * GHC.Nat-indexed lists, originally by Tom Smeding
+
+-- This needs to be moved to a better place, probably its own module,
+-- once its clear which other modules depend on it.
+
+-- | An index in an n-dimensional array. The fastest-moving index is at the
+-- last position; thus the index 'Z :. i :. j' represents 'a[i][j]' in
+-- traditional C notation.
+data Index (n :: Nat) i where
+  Z :: Index 0 i
+  S :: i -> Index n i -> Index (n + 1) i
+
+deriving stock instance Functor (Index n)
+
+instance Show i => Show (Index n i) where
+  showsPrec _ Z = showString "Z"
+  showsPrec d (idx :. i) = showParen (d > 3) $
+    showsPrec 3 idx . showString " :. " . showsPrec 4 i
+
+-- I'm afraid, I can't do the unsafeCoerces below with this order
+-- of argument in :., can I? So I need this instead:
+pattern (:.) is i = S i is
+infixl 3 :.
+-- This fixity is stolen from Accelerate:
+--   https://hackage.haskell.org/package/accelerate-1.3.0.0/docs/Data-Array-Accelerate.html#t::.
+
+tailIndex :: Index (1 + n) i -> Index n i
+tailIndex (is :. i) = is
+
+takeIndex :: forall len n i. KnownNat len
+          => Index (len + n) i -> Index n i
+takeIndex ix = unsafeCoerce $ take (valueOf @len) $ unsafeCoerce ix
+
+dropIndex :: forall len n i. KnownNat len
+          => Index (len + n) i -> Index n i
+dropIndex ix = unsafeCoerce $ drop (valueOf @len) $ unsafeCoerce ix
+
+permutePrefixIndex :: [Int] -> Index n i -> Index n i
+permutePrefixIndex p ix =
+  let l = unsafeCoerce l
+  in unsafeCoerce $ V.toList $ VS.fromList l V.// zip p l
+
+-- | The shape of an n-dimensional array. Represented by an index to not
+-- duplicate representations and convert easily between each. It seems unlikely
+-- enough to make mistakes even with this dumb wrapper, so it might be fine.
+newtype Shape n i = Shape (Index n i)
+  deriving Show
+
+-- It seems that function names can't start with colon. That's too bad.
+-- Also, I can't make pattern synonym of out that because newtype is in the way.
+-- Or can I if I defined two pattern synonyms?
+infixl 3 @$
+(@$) :: Shape n i -> i -> Shape (1 + n) i
+Shape sh @$ s = Shape (sh :. s)
```

Comment:
@tomsmeding: this is the second.

Reply:
View patterns and explicitly bidirectional pattern synonyms are way too powerful.
```haskell
pattern (::.) :: Shape n i -> i -> Shape (1 + n) i
pattern sh ::. i <- ((\(Shape (sh :. i)) -> (Shape sh, i)) -> (sh, i))
  where pattern (Shape sh) ::. i = Shape (sh :. i)
```
I hope I did that correctly, don't have a compiler handy.

Reply:
With the simplified (:.) that is no longer a pattern synonym, I've almost made it to work and it's so useful and improves safety a lot. I think the only wrong detail is the type of the new pattern, which then causes these to fail:

```
  AstTranspose v -> case shapeAst v of
    Shape (i :. k :. sh) -> k :$ i :$ Shape sh
    sh -> sh  -- the operation is identity if rank too small

-- even this fails (probably the type of (:$) is not hairy enough):
--    ZS -> ZS
--    a :$ bs -> a :$ bs

-- this is how it should look:
--    i :$ k :$ sh -> k :$ i :$ sh
--    sh -> sh  -- the operation is identity if rank too small
```

Reply:
This seems to work? Up to you to fix code style and organisation https://paste.tomsmeding.com/IKwLCzCk

Reply:
Mind blown. Thank you.

Reply:
The thing is that your type was correct and mine was not (indeed `n` should be existential, and the equality constraint should be provided -- this is precisely what you wrote). However, the code didn't typecheck under this correct type signature because the type of that inline lambda would be very weird: it would return something of existential type, which is dodgy.

So I "simply" added an explicit type signature to that lambda (and extracted it to a top-level function for sanity reasons), encoding the existential in a data type; and then we still need to export the proof that relates the existential `n` to the original `n1`, so we export that too. Then you get this code.

Reply:
(My knowledge comes from writing an early incorrect version of, and later understanding the actually correct version of, [this code here](https://github.com/AccelerateHS/accelerate/blob/5971c5d8e4dbba28d2017e7ce422cf46a20197cb/src/Data/Array/Accelerate/AST/Idx.hs#L65-L92))

Reply:
"Simply" is well put.